discovery_engine_core/core/src/engine.rs

```diff
@@ -874,185 +1049,60 @@ fn documentify_articles(
-        let stack_ops = vec![
-            Box::new(BreakingNews::new(
-                &endpoint_config,
-                providers.headlines.clone(),
-            )) as BoxedOps,
-            Box::new(TrustedNews::new(
-                &endpoint_config,
-                providers.trusted_headlines.clone(),
-            )) as BoxedOps,
-            Box::new(PersonalizedNews::new(
-                &endpoint_config,
-                providers.news.clone(),
-            )) as BoxedOps,
-        ];
-
-        let (mut stack_data, builder) = if let Some(state) = state {
-            if stack_ops.is_empty() {
-                return Err(Error::NoStackOps);
-            }
-            State::deserialize(state, builder)?
-        } else {
-            (HashMap::default(), builder)
-        };
-        for id in stack_ops.iter().map(Ops::id).chain(once(Exploration::id())) {
-            if let Ok(alpha) = de_config.extract_inner::<f32>(&format!("stacks.{id}.alpha")) {
-                stack_data.entry(id).or_default().alpha = alpha;
-            }
-            if let Ok(beta) = de_config.extract_inner::<f32>(&format!("stacks.{id}.beta")) {
-                stack_data.entry(id).or_default().beta = beta;
-            }
-        }
-
-        let ranker = builder.build()?;
-        #[cfg(feature = "storage")]
-        let storage = {
-            let storage = SqliteStorage::connect("sqlite::memory:").await?;
-            storage.init_database().await?;
-            Box::new(storage) as _
-        };
-
-        Self::new(
-            endpoint_config,
-            core_config,
-            exploration_config,
-            ranker,
-            history,
-            sources,
-            stack_data,
-            stack_ops,
-            #[cfg(feature = "storage")]
-            storage,
-            providers,
-        )
-        .await
-    }
-}
-
 #[derive(Serialize, Deserialize)]
-struct StackState(Vec<u8>);
+struct SerializedStackState(Vec<u8>);
 
-impl StackState {
+impl SerializedStackState {
     fn deserialize(&self) -> Result<HashMap<StackId, StackData>, Error> {
         bincode::deserialize(&self.0).map_err(Error::Deserialization)
     }
 }
 
 #[derive(Serialize, Deserialize)]
-struct RankerState(Vec<u8>);
+struct SerializedCoiSystemState(Vec<u8>);
 
-impl RankerState {
-    fn deserialize<'a>(
-        &self,
-        builder: Builder<'a, AveragePooler>,
-    ) -> Result<Builder<'a, AveragePooler>, Error> {
-        builder.with_serialized_state(&self.0).map_err(Into::into)
+impl SerializedCoiSystemState {
+    fn deserialize(&self) -> Result<CoiSystemState, Error> {
+        CoiSystemState::deserialize(&self.0).map_err(Into::into)
     }
 }
 
 #[derive(Serialize, Deserialize)]
-struct State {
+struct SerializedState {
     /// The serialized engine state.
-    engine: StackState,
+    stacks: SerializedStackState,
     /// The serialized ranker state.
-    ranker: RankerState,
+    coi: SerializedCoiSystemState,
```

Comment:
:grey_question: do the associated doc strings need to be updated? e.g. the serialized stack state